### ISSUE TITLE
Allow different storage backends for biicode

### ISSUE THREAD

**Issue by drodri** [biicode/biicode#6]
Now, just the biicode backend is available, due to historic reasons. Though git is the most reasonable one, many users still use svn and others, so we just went for the most simple way: just use the biicode storage. As all the metainformation for files was storing, just storing the contents/blobs for the files was really straightforward.

This can make challenging to to implement this issue, and I think some thinking and planning is required. I will first analyse things, current status, interactions, processes and storage and retrieval, but a possible outline of the process in order to incrementally build it without breaking things, could be:

- Decouple retrieval of meta information from blobs retrieval. For dependencies it can be checked here: https://github.com/biicode/common/blob/develop/deps/closure_builder.py. No new functionality here.
- Introduce a level of indirection in the biicode blobs that will redirect to the real backend. No new functionality here, just running with biicode backend.
- Define syntax in biicode.conf for different backends. (maybe starting with git, then zip, http... in second iteration). Just allowing the syntax.
- Specifics of backends.

Ross @rharriso offered (thanks!) collaboration with this issue. Check if we are talking about the same thing and please let me check everything, and will come back to this issue, refine plan, and lets talk about concrete tasks and next steps.

**Comment by rharriso** [comment]
I'm still familiarizing myself with the source code. But I have some questions already.

* what is included in the meta data? I am assuming that blob retrieval refers to fetching of block code.
    * Does the metadata include build information, and location of required headers & objects?
* Would these changes require an entry in the biicode repository? 
    * This could be a way to prevent circumvention of the [premium features](http://biicode-docs.readthedocs.org/c++/faqs.html#is-biicode-free).
* Is there a test suite? Or development documentation?

**Comment by drodri** [comment]
Yes, metadata basically consist of:

Cell:
- Name & ID
- Declaration of dependencies (#includes, can be resolved or not)
- Dependencies between files (E=Explicit, coming from resolved #includes, I=Implicit, coming for implementations and manual configuration).
- Necessary paths to resolve depedencies (will be translated to include-paths)
- If the file contains a main()

Content:
- ID
- Blob of data (lazily compressed) with its sha
- Parsing info: symbols (classes, methods), #includes, that will generate Declarations

Both Cell and Content form a Resource.
When things are retrieved from the biicode repository, are retrieved as Resources, and that is the problem: it has to be decoupled to first retrieve the meta-information of Cells & Contents, but without the actual blobs of data, so they can be directly retrieved from other place.

About circumvention of premium features, it is always possible, but not simple. Still a biicode server is needed as the meta information is stored there. As the circumvention of premium features, it is not a big concern right now, we just want to create something really useful that solves our problems.

There are some tests within the repos, but those are just a part. We still have many more in a separate private repo, as it is also highly coupled with our backend and sensitive data. We still have tons of work to do to decouple things, so we will be continously porting tests from there to the client and common repos. Unfortunately there is no development documentation, we are small team and work a lot on the board/paper in the office, and the design/implementation has evolved wildly and very fast, so we didnt have the need and time for development documentation, though we have started some time ago to be more thorough with our docstrings, but many still to do here....

**Comment by drodri** [comment]
Hi @rharriso
I have started to take small steps, cleaning things and studying how to do it, I explain them here just in case you want to track changes, and what is being done.
- First thing was a bit of cleaning in Blobs storage for bytes, and avoid serializating them into our DB for "opened" blocks: https://github.com/biicode/common/pull/5
- Then, the dependencies of a given project were also serialized to the project DB, which it is useless, as they are computed again: https://github.com/biicode/common/pull/7
Still some work to prepare things, and still not clear how to approach the problem, lets keep decoupling things.

**Comment by rharriso** [comment]
To handle decoupling. How should we handle notation.
* tags use @tag.number
* perhaps we could use git:git-pull-path, local::path/to/local/folder ( those two examples would be most useful to me )

About your changes:
This question will reveal my ignorance. How are dependencies for blocks stored if not in the local database?

**Comment by drodri** [comment]
I was talking about the "opened" blocks, not the dependencies. The opened blocks are those that are actually being edited, typically it is just one, but you can have many of them open at the same time.
For opened blocks, lots of unnecessary meta-information was being stored to DB, and it was unnecessary as it is computed again (and faster) when a new invocation of "bii" is done.

Another small step has been done, with another large improvement in processing speed, especially for large projects: https://github.com/biicode/common/pull/9

Also, the dependencies and their meta-information are stored in the biicode user cache, common for all projects and located in your .biicode folder, but they were also stored (again) in the project DB, which was redundant and slow, and that is what has been removed.

Note that still much work has to be done until we have to actually specify the notation for referring to git and paths. Still not sure about how to handle the decoupling, at least these small steps are clearing the way and simplifying the codebase. I wanted to ask for more input: how do you envision the retrieval from git? because now, the meta-information for files and blocks is stored in biicode DB, so actually referring to git to just retrieve the files has little gain (it is not that complex to upload the files to biicode, as we do just now). If we commit the meta-information into text or binary files into the git repo, we have to design a full new system to retrieve and handle meta-information from there, and that will not be simple. 
Thanks! Cheers.

**Comment by rharriso** [comment]
Answering your last point. The more I think about it, there isn't much gain. Either way there will have to be a definition on the server of how to build and include the block. 

What I've used this for with bundler was working on my own forks (locally then on git) before committing to the root project. However, given the different work environments, the best course might be to encourage projects to use biicode, and build my own forks locally.

I still think there is utility in this, but not nearly as immediate as I once thought.

**Comment by drodri** [comment]
Yes, that is the point. There is a good point if we are able to store meta-information in plain text files (and manage it coherently, which is a challenge) and also interact directly with git (and other stores), so yes, you could interact directly with your local copy. Not sure how we could manage the names and version, but certainly it can be done, but could require a lot of work, and now there are many other inmediate priorities. That is why I have started a roadmap for this decoupling, and we address it when we can find some time. I would say that in about 3 months it will be ready for working on this feature and providing alternative origins. Lets keep working on it. Cheers.

**Comment by drodri** [comment]
Similar to https://github.com/biicode/client/issues/23
